Instrument editor love

charbeljc opened the pull request:
Ok, second batch for the instrument editor, which become finally mostly usable :grin: 

On the technical side, wave widgets now track mouse movements continuously, this allow us to highlight envelope points before we drag / delete them, and see why editing start/loop/end frames in main widget is broken (final PR to come :grinning: )

@theGreatWhiteShark if you could have a look ...
Regards.

charbeljc commented:
I have one more commit pending before this pull request is complete ...

charbeljc commented:
Ok, I'm done with Sample Editing :grinning:

theGreatWhiteShark commented:
> @theGreatWhiteShark if you could have a look ...
>
> Ok, I'm done with Sample Editing grinning

That's quite fast :smile: . I'll try to take a look within the next days.

theGreatWhiteShark commented:
Hey @charbeljc ,

As these are improvements rather than bug fixes could you change the target to branch `development`? (Sorry for the inconvenience. This will be straight more straight forward once 1.1 is release and we use `master` for incremental improvements again)

You do not need to close this one and open a new one but can just press the "Edit" button to the right of the title of this PR and select a different branch via a combo box.

theGreatWhiteShark reviewed (commented):
Thanks for your contribution! I like it a lot!

Could you also have a look at the coding conventions in the  [DEVELOPERS](https://github.com/hydrogen-music/hydrogen/blob/master/DEVELOPERS) file in our repo? I know that our code right now is quite a mess. But if we all stick to the same conventions it will become more easily readable in the future.

I also found two further inconsistencies in the Sample Editor (which are probably not related to your PR)
- open the SampleEditor -> enter a large value in the spinbox of the "Start" and press enter -> both the values in the spinboxes of "Start" and "Loop" get updated but only the "S" line gets moved in the wave display while "L" remains at zero.
- right to the "Play original sample" button there is a the label "new sample length" which seem to show incorrect values until the "E" marker is moved.

charbeljc commented on src/gui/src/SampleEditor/TargetWaveDisplay.cpp:
```diff
@@ -238,117 +240,123 @@ void TargetWaveDisplay::updateDisplay( H2Core::InstrumentLayer *pLayer )
+			// if only 2 points, remove them both
+			envelope.clear();
+		} else {
+			envelope.erase( envelope.begin() +  m_nSelectedEnvelopePoint );
```
Good catch :grin:, see 29edd84e5db4f6fcaa91ed821d2d84b4d8e915f7

charbeljc commented on src/gui/src/SampleEditor/SampleEditor.h:
```diff
@@ -55,9 +55,10 @@ class SampleEditor : public QDialog, public Ui_SampleEditor_UI, public H2Core::O
+		void setUnclean();
+		void setClean();
+
+		bool m_bSampleEditorClean;
```
done, see 397034a258a43d8fa562b62bf388e7dd02a5b0c1

charbeljc commented on src/gui/src/SampleEditor/SampleEditor.cpp:
```diff
@@ -345,8 +344,10 @@ void SampleEditor::on_PrevChangesPushButton_clicked()
+	setClean();
 	QApplication::restoreOverrideCursor();
+	InstrumentEditorPanel *panel = InstrumentEditorPanel::get_instance();
+	panel->selectLayer( panel->getSelectedLayer() ); // reselect layer to trigger update
```
see 499fbca0a9972a2176653d61994c9835b1077fd0

charbeljc commented on src/gui/src/SampleEditor/MainSampleWaveDisplay.h:
```diff
@@ -50,17 +58,20 @@ class MainSampleWaveDisplay : public QWidget, public H2Core::Object
 		bool	m_bLoopSliderIsMoved;
 		bool	m_bEndSliderIsmoved;
 
+		Slider  m_SelectedSlider;
```
done, see 710676298b659cb52a758e3926759870758b629d

charbeljc commented on src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp:
```diff
@@ -283,3 +303,26 @@ void MainSampleWaveDisplay::mouseReleaseEvent(QMouseEvent *ev)
 
 
+
+void MainSampleWaveDisplay::chooseSlider(QMouseEvent * ev)
```
For the y target coordinates, I choose mostly the y location of the slider's labels (S, L, E), also, the slider selection is done on mouse press and never change while dragging.

Before, the selected slider could change while dragging, furthermore, there was a dead zone between each slider influence.

The previous behavior  could you drive nuts ...

theGreatWhiteShark replied:
Definitely. It's a huge improvement. The only thing I found a little bit strange was that clicking in the space between two adjacent sliders in order to move one of them there yields different results depending on the y coordinate of the mouse pointer.

charbeljc replied:
> Definitely. It's a huge improvement. The only thing I found a little bit strange was that clicking in the space between two adjacent sliders in order to move one of them there yields different results depending on the y coordinate of the mouse pointer.

Yeah, but I think we can get used of this behavior. At least we can make a mental model from it, and having the actual slider selected highlighted helps a lot.

charbeljc commented on src/gui/src/SampleEditor/TargetWaveDisplay.cpp:
```diff
@@ -80,12 +84,31 @@ TargetWaveDisplay::~TargetWaveDisplay()
+		painter.setPen( QPen(lineColor, 1 , Qt::SolidLine) );
+		painter.drawLine( envelope[i]->frame, envelope[i]->value, envelope[i + 1]->frame, envelope[i +1]->value );
+		if ( i == selected )
+			painter.setBrush( selectedColor );
```
done, see 2b6ff4cf237317c3c46c68707dc1990318cd0a03

charbeljc commented on src/gui/src/SampleEditor/TargetWaveDisplay.cpp:
```diff
@@ -238,117 +240,123 @@ void TargetWaveDisplay::updateDisplay( H2Core::InstrumentLayer *pLayer )
+			update_envelope(envelope, m_nSelectedEnvelopePoint, m_nX, m_nY, m_nSnapRadius);
+		}
+	} else if (ev->button() == Qt::RightButton ) {
+		//remove point
```
done also, see 2b6ff4cf237317c3c46c68707dc1990318cd0a03

charbeljc commented on src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp:
```diff
@@ -228,53 +267,34 @@ void MainSampleWaveDisplay::testPosition( QMouseEvent *ev )
 	}
 
 //loopframeposition
-	else if  (ev->y()<=65 ) {
```
Yes, this was for consistency with the spinbox behavior. When you try to decrease loop_frame before start_frame, the spinbox clip it at start_frame. Let me know what you think about it.

theGreatWhiteShark replied:
Ah, I see. I think it's fine.

charbeljc commented on src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp:
```diff
@@ -200,25 +217,47 @@ void MainSampleWaveDisplay::updateDisplay( const QString& filename )
+		chooseSlider( ev );
+	}
 	update();
+	if (propagate()){
```
Hum, you are right ...

charbeljc replied:
see cb9866a7ff725a418435133b6480d067d266bae4

charbeljc commented:
@theGreatWhiteShark Thanks for your review, I addressed most of your comments and I should be really done this time.
Let me know what you think and I will rebase this PR on development branch. Regards,

charbeljc commented:
> Thanks for your contribution! I like it a lot!
> 
> Could you also have a look at the coding conventions in the [DEVELOPERS](https://github.com/hydrogen-music/hydrogen/blob/master/DEVELOPERS) file in our repo? I know that our code right now is quite a mess. But if we all stick to the same conventions it will become more easily readable in the future.
> 
> I also found two further inconsistencies in the Sample Editor (which are probably not related to your PR)
> 
>     * open the SampleEditor -> enter a large value in the spinbox of the "Start" and press enter -> both the values in the spinboxes of "Start" and "Loop" get updated but only the "S" line gets moved in the wave display while "L" remains at zero.
> 
>     * right to the "Play original sample" button there is a the label "new sample length" which seem to show incorrect values until the "E" marker is moved.

I'll have a look at this later ! (and also, I will try to follow developper's style guide)

theGreatWhiteShark commented on src/gui/src/SampleEditor/MainSampleWaveDisplay.cpp:
```diff
@@ -199,26 +218,41 @@ void MainSampleWaveDisplay::updateDisplay( const QString& filename )
+		chooseSlider( ev );
+	}
 	update();
+	mouseUpdateDone();
```
How come you mark the state of the SampleEditor unclean once the user enters the MainSampleWaveDisplay using the mouse? This makes the "Apply Changes" not flat and indicates something has changed. Changes via drag should be already covered by the call of `mouseUpdateDone` in `mousePressEvent`

charbeljc replied:
> How come you mark the state of the SampleEditor unclean once the user enters the MainSampleWaveDisplay using the mouse? This makes the "Apply Changes" not flat and indicates something has changed. Changes via drag should be already covered by the call of `mouseUpdateDone` in `mousePressEvent`

Hum, actually, it does not mark the editor as changed, only setClean()/setUnclean() does that. I should maybe rename `mouseUpdateDone()` to something like `mouseActionDone()`

theGreatWhiteShark reviewed (commented):
Nice.

There are still some oddities but they were most probably already present so you do not need to fix them.

- the aforementioned: open the SampleEditor -> enter a large value in the spinbox of the "Start" and press enter -> both the values in the spinboxes of "Start" and "Loop" get updated but only the "S" line gets moved in the wave display while "L" remains at zero.
- Open the SampleEditor -> change something and hit the "Close" button -> a dialog "Unsave changes..." will pop up and the user has to consent to discard the changes -> clicking "Ok" and reopen the SampleEditor by clicking "Edit Layer" again -> the same popup shows up again. I think the second one is a dud.
- When entering a fresh instance of the SampleEditor the value of the `EndFrameSpinBox` looks perfectly fine. But as soon as the mouse pointer enters the `MainSampleWaveDisplay` it is reset to 0 by `SampleEditor::getAllFrameInfos()` what causes some other oddities. The slider, however, stays at it's correct position.

theGreatWhiteShark commented:
In addition, could you add all the changes in the UX to the [ChangeLog](https://github.com/hydrogen-music/hydrogen/blob/development/ChangeLog) once you rebased on `development`?

charbeljc commented:
> In addition, could you add all the changes in the UX to the [ChangeLog](https://github.com/hydrogen-music/hydrogen/blob/development/ChangeLog) once you rebased on `development`?

Ahem, the ChangeLog... Those are my weak point... :smile:

I just rebased this PR on development branch, see #1302. I would be glad if it could be merged soon, because I have interesting things to come, namely python bindings for the core library, UI as a shared library, with python bindings also, and it starts to make quite a big stack of commits :grin:

theGreatWhiteShark commented:
> I just rebased this PR on development branch, see #1302. I would be glad if it could be merged soon, because I have interesting things to come,

Sure.

> See #1303 for the bindings grinning

Nice!
